### PR TITLE
PFW-1024

### PR DESCRIPTION
**Pull request by leptun**
Backlight control implementation. Work in progress. It dims the backlight automatically after 10s.
After it is complete, the user will be able to set backlight brightness from the settings menu and adjust the timeout for reduced brightness.

### CODE REVIEW

**Comment by 3d-gussner** [comment]
@leptun Maybe stupid question but looking into the [reprap LCD screen schematics](https://reprap.org/mediawiki/images/7/70/Controller_final_reprapdiscount.pdf) the [pin you use](https://github.com/ultimachine/Einsy-Rambo/blob/1.1a/board/Project%20Outputs/Schematic%20Prints_Einsy%20Rambo_1.1a.PDF) to dim isn't connected to anything. Or does the Prusa LCD display has that pin connected? Or does it need a hardware hack?

**Comment by jakk2809** [comment]
I was about to add a pot and switch for the LCD backlight (inspired by https://www.youtube.com/watch?v=ymwqs_njZbk), but found that the connection to the backlight is not that simple. It looks like PE3 (PWM5) might have the ability to control the brightness. 

I think the schematic is as follows:
IMG_0065

IMG_0022

I guess I could just take everything out and still wire in the pot and switch, but if it can be controlled through the software (and the ability to save the brightness setting), that would be nice, too. I have a recent MK3S (September 2019) with a LCD 2004A-V3.0 and LDO-RRLCD-A004.

**Comment by leptun** [comment]
@jakk2809 I think that this is the right schematic. I don’t have the schematic, but it works correctly if I try to control PE3. (Edit)

@3d-gussner The Prusa lcd on MK3(S) is simmilar to the reprap one, but it has some extra circuitry for brightness control. This display is owned by (I think) most, if not all MK3(S) users. 

The PR is not ready yet. There is still some polishing to do for the UI.

**Comment by jakk2809** [comment]
@leptun Thank you for working on this. 

I couldn't find a schematic of the most recent revision online so I just probed around. Also, I think I miss-labeled R9 and R8 on my schematic.

**Comment by 3d-gussner** [comment]
@DRracer Thanks, so the manual pictures for the MK3s are old :-) https://manual.prusa3d.com/Guide/6.+LCD+assembly/1050?lang=en#s18411

**Comment by leptun** [comment]
I am testing right now if I can detect from the code if backlight is present. Can you also test on your printer with lcd with and without backlight and see if it detects it (option is missing in settings if backlight is not detected). In my testing it detected accordingly, but I want to make sure it’s the same for everyone. You must have an einsy1.1a/b board for this to work. On older boards with the SPI daughterboard, the backlight pin is used for accessing the external flash and the pin used on newer boards is just left unconnected on the board.

**Comment by PaulGD03** [comment]
Hi there! I love your idea of dimming the backlight, which enables me to *finally* throw away my printed lcd cover! ^^ I have personally added a "Night mode" which dims it down to a value of 5, and I'm currently looking for a place that's best suited for these settings... HW Setup maybe? 

I've attached two images below showing the night mode, maybe it's something that could be considered to be added?

**Comment by leptun** [comment]
Since it’s not finalised, it has no place for now. There will probably be a “Backlight” submenu in settings where you have normal level, dimmed level, dimming timeout (0 to disable). I find such night mode quite redundant if you can set the low brightness manually. I’ll keep making changes so stay tuned. ;)

**Comment by leptun** [comment]
Btw, what’s up with your status screen? Is that some debugging or is that corruption?

**Comment by PaulGD03** [comment]
> Since it’s not finalised, it has no place for now. There will probably be a “Backlight” submenu in settings where you have normal level, dimmed level, dimming timeout (0 to disable). I find such night mode quite redundant if you can set the low brightness manually. I’ll keep making changes so stay tuned. ;)

Yea makes sense, that's actually what I'm trying to do for myself now, I'm impatient lol :P  

> Btw, what’s up with your status screen? Is that some debugging or is that corruption?

Honestly? No idea! That's just how it shows up when I compiled and flashed the MK3 branch with your commits added... I'll try with the 3.8.1-RC1 Release code, maybe something got messed up with the lcd again haha.

**Comment by leptun** [comment]
If you are using arduino ide (but not limited to it) you have to change some platform changes to enable full printf support. Follow the build instructions carefully and make sure to set the compiler flags right in the config.

**Comment by PaulGD03** [comment]
> If you are using arduino ide (but not limited to it) you have to change some platform changes to enable full printf support. Follow the build instructions carefully and make sure to set the compiler flags right in the config.

Ah alright, I guess I didn't read it well enough and missed something haha. Will do, hopefully that way my eyes will hurt less while I experiment with this a bit more ^^

**Comment by Panayiotis-git** [comment]
For the second line's "Q!0", the PLANNER_DIAGNOSTICS is defined.
For the third line's "C04", the CMD_DIAGNOSTICS is defined.

**Comment by leptun** [comment]
UI changes incoming... :)

I've made some significant changes to the way the backlight implementation works under the hood.
I've also moved all backlight related stuff in a submenu named "Brightness". Please note that the changes are saved to eeprom AFTER you exit the submenu. Don't expect to make a brightness change, immediately reset the printer and wonder why it didn't save.
In that submenu you'll find a way to adjust both HIGH and LOW brightness settings and also a toggle for dimming type (Always on, Always off, Auto). I'm planning to add another option for setting the timeout in seconds.
Please let me know how you feel with the current UI. Is there anything you'd like to change?

Edit: Here's a video with a showcase: https://photos.app.goo.gl/WUmDmkBvdAsjxdfJ9
The flickering is only on camera ;)

**Comment by 3d-gussner** [comment]
@leptun looks really good...have to test it myself

**Comment by PaulGD03** [comment]
@leptun Nice one, looks really clean! Will test it soon on my own printer ^^

**Comment by leptun** [comment]
@3d-gussner @PaulGameDev @jakk2809 
Now we also have adjustable timeout ;)
If anyone can test this on their printer, I would be greatful. Keep in mind that if you used the version without variable timeout, then the default value will be -1s. It can easily be adjusted, just a random fact. For anyone using this PR for the first time, the issue is non-existent.

I'll continue by adding automatic wake triggers on some of the notifications.
Edit: Also you can no longer have inverted brightness levels. Bright is always at least as bright as Dim and the other way around.

**Comment by PaulGD03** [comment]
Awesome! Been some time since I've recompiled my firmware so often lol

**Comment by 3d-gussner** [comment]
I got an old  LCD screen and can't test it :-(

**Comment by PaulGD03** [comment]
> I got an old LCD screen and can't test it :-(

Aw man, that must suck! :/

Well, I can at least confirm that it does work correctly on mine! Nice job leptun ^^

**Comment by leptun** [comment]
Update:
I've added the option to change the brightness in the Tune menu.
During PowerPanic the backlight is turned off now in order to preserve even more power.

**Comment by vintagepc** [comment]
Oooh, nice feature. Are you in need of additional testers? I'm curious how this will work on my modded orange screen... (OEM board, just the LCD module replaced)

image

**Comment by leptun** [comment]
@vintagepc Sure, testers are welcome. Right now I need to find edge cases where the backlight misbehaves. I’m also doing some changes to the UI in regard to triggers for waking the display. I still have to do some changes since I didn’t finish everything last night.

Btw, does your lcd assembly contain the backlight circuitry required for controlling the brightness with Exp2 pin1?
You can just try using my PR and see if the settings appear in the menus and backlight changes automatically. If the above doesn’t happen, then you have the old display.

**Comment by vintagepc** [comment]
Well that came to a screeching halt. My Mk3 is fairly new (Jan 2019) and an even newer second display module (May or so?) when the first one developed flaky characters. 

Both are LDO-RRLCD-A003 and do not have the control circuitry. 

This does not bode well for the number of folks that could actually make use of this feature... it would appear it exists only on A004 and newer.

**Comment by leptun** [comment]
@jakk2809 @PaulGameDev When did you get your printer/display? You are the only ones on this PR to have backlight controllable LCDs (+me).

**Comment by jakk2809** [comment]
I got my printer in September 2019. LCD 2004A-V3.0 and LDO-RRLCD-A004.

**Comment by vintagepc** [comment]
Minor correction - I typo'd the date, I got my printer Jan 2019, not 2018. (Fixed original post)

**Comment by PaulGD03** [comment]
@leptun I got mine in June 2019. Will check for the writings on the LCD later today

**Comment by leptun** [comment]
@DRracer Updated toggle.

File: Firmware/backlight.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
+//backlight.h
+#ifndef _BACKLIGHT_H
+#define _BACKLIGHT_H
+
+#include <inttypes.h>
+#include "Marlin.h"
+#include "pins.h"
+
+enum Backlight_Mode
```
Comment (DRracer) [inline Firmware/backlight.h:9]: we can probably save a few bytes making this an enum class of type uint8_t
the compiler won't expect these numbers being 16bit integers then

**Review by DRracer — APPROVED** [review]

**Pull request merged**